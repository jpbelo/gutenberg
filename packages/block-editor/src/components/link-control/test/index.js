/**
 * External dependencies
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act, Simulate } from 'react-dom/test-utils';

/**
 * Internal dependencies
 */
import LinkControl from '../index';
import { fauxEntitySuggestions, fetchFauxEntitySuggestions } from './fixtures';

function currentEventLoopEnd() {
	return new Promise( ( resolve ) => setImmediate( resolve ) );
}

let container = null;
beforeEach( () => {
	// setup a DOM element as a render target
	container = document.createElement( 'div' );
	document.body.appendChild( container );
} );

afterEach( () => {
	// cleanup on exiting
	unmountComponentAtNode( container );
	container.remove();
	container = null;
} );

describe( 'Basic rendering', () => {
	it( 'should render with required props', () => {
		act( () => {
			render(
				<LinkControl
				/>, container
			);
		} );

		// Search Input UI
		// const searchInputLabel = Array.from( container.querySelectorAll( 'label' ) ).find( ( label ) => label.innerText === 'Search or input url' );
		const searchInput = container.querySelector( 'input[aria-label="URL"]' );

		// expect( searchInputLabel ).not.toBeNull();
		expect( searchInput ).not.toBeNull();

		expect( container.innerHTML ).toMatchSnapshot();
	} );
} );

describe( 'Searching', () => {
	it( 'should render search results for current search term', async ( ) => {
		const searchTerm = 'Hello';

		act( () => {
			render(
				<LinkControl
					fetchSearchSuggestions={ fetchFauxEntitySuggestions }
				/>, container
			);
		} );

		let searchResultElements;

		// Search Input UI
		const searchInput = container.querySelector( 'input[aria-label="URL"]' );

		// TODO: select these by aria relationship to autocomplete rather than arbitary selector.
		searchResultElements = container.querySelectorAll( '[role="menu"] button[role="menuitem"]' );

		expect( searchResultElements ).toHaveLength( 0 );

		// Simulate searching for a term
		act( () => {
			Simulate.change( searchInput, { target: { value: searchTerm } } );
		} );

		// fetchFauxEntitySuggestions resolves on next "tick" of event loop
		await currentEventLoopEnd();

		// TODO: select these by aria relationship to autocomplete rather than arbitary selector.
		searchResultElements = container.querySelectorAll( '[role="listbox"] button[role="option"]' );

		expect( searchResultElements ).toHaveLength( fauxEntitySuggestions.length );

		// Reset the search term
		act( () => {
			Simulate.change( searchInput, { target: { value: '' } } );
		} );

		// fetchFauxEntitySuggestions resolves on next "tick" of event loop
		await currentEventLoopEnd();

		// TODO: select these by aria relationship to autocomplete rather than arbitary selector.
		searchResultElements = container.querySelectorAll( '[role="listbox"] button[role="option"]' );

		expect( searchResultElements ).toHaveLength( 0 );
	} );
} );
