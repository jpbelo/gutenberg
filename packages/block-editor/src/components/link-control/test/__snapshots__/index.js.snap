// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic rendering should render with required props 1`] = `"<span><div tabindex=\\"-1\\"><div><div><div class=\\"components-popover editor-url-popover block-editor-url-popover is-bottom is-center components-animate__appear is-from-top\\" style=\\"\\"><div class=\\"components-popover__content\\" tabindex=\\"-1\\"><div class=\\"block-editor-url-popover__input-container\\"><div class=\\"editor-url-popover__row block-editor-url-popover__row\\"><div class=\\"block-editor-link-control__popover-inner\\"><div class=\\"block-editor-link-control__search\\"><form><div class=\\"editor-url-input block-editor-url-input block-editor-link-control__search-input\\"><input type=\\"text\\" aria-label=\\"URL\\" required=\\"\\" placeholder=\\"Search or type url\\" role=\\"combobox\\" aria-expanded=\\"false\\" aria-autocomplete=\\"list\\" aria-owns=\\"block-editor-url-input-suggestions-0\\" value=\\"\\"></div></form></div></div></div></div></div></div></div></div></div></span>"`;
